src/greeter/waylandkeyboardbackend.cpp

```diff
@@ -35,14 +43,55 @@ WaylandKeyboardBackend::~WaylandKeyboardBackend()
 {
 }
 
-void WaylandKeyboardBackend::init()
+
+QList<QObject *> parseRules(const QString &filename, int &current)
 {
-    d->layouts.clear();
+    QFile file(filename);
+    qDebug() << "Parsing xkb rules from" << file.fileName();
+    if (!file.open(QFile::ReadOnly | QFile::Text)) {
+        qWarning() << "Cannot open the rules file" << file.fileName();
+        return {};
+    }
+
+    QList<QObject *> layouts;
+
+    QString lastName, lastDescription;
+
+    QStringList path;
+    QXmlStreamReader reader(&file);
+    while (!reader.atEnd()) {
+        const auto token = reader.readNext();
+        if (token == QXmlStreamReader::StartElement) {
+            path << reader.name().toString();
+            QString strPath = path.join(QLatin1String("/"));
 
-    QDir dir(QStringLiteral("/usr/share/X11/xkb/symbols"));
-    auto entries = dir.entryList(QDir::Files);
-    for (const auto &entry : qAsConst(entries))
-        d->layouts << new KeyboardLayout(entry, entry);
+            if (strPath.endsWith(QLatin1String("layoutList/layout/configItem/name"))) {
+                lastName = reader.readElementText().trimmed();
+            } else if (strPath.endsWith(QLatin1String("layoutList/layout/configItem/description"))) {
+                // TODO: This should be translated using i18nd("xkeyboard-config", lastDescription)
+                lastDescription = reader.readElementText().trimmed();
+            }
+        }
+        // don't use token here, readElementText() above can have moved us forward meanwhile
+        if (reader.tokenType() == QXmlStreamReader::EndElement) {
+            const QString strPath = path.join(QLatin1String("/"));
+            if (strPath.endsWith(QLatin1String("layoutList/layout/configItem/description"))) {
+                layouts << new KeyboardLayout(lastName, lastDescription);
+            }
+            path.removeLast();
+        }
+    }
+
+    if (reader.hasError()) {
+        qWarning() << "Failed to parse the rules file" << file.fileName();
+        return {};
+    }
+    return layouts;
+}
+
+void WaylandKeyboardBackend::init()
+{
+    d->layouts = parseRules(QStringLiteral("/usr/share/X11/xkb/rules/evdev.xml"), d->layout_id);
```

Comment:
`libxkbregistry` is upstream-blessed library for parsing such things. We could use it instead of parsing `evdev.xml` directly:
> The libxkbregistry API to query for available rules, models, layouts, variants and options (RMLVO).
>
> libxkbregistry is a separate library to libxkbcommon.
>
> **This library is the replacement for clients currently parsing evdev.xml directly.** The library is intended to provide easy access to the set of possible MLVO configurations for a given ruleset. It is not a library to apply these configurations, merely to enumerate them. The intended users of this library are the configuration UIs that allow a user to select their keyboard layout of choice. 

https://xkbcommon.org/doc/current/group__registry.html

Reply:
Do you know if it does the translating or we still need to do it ourselves?

TBF, this code already works (as it's taken from KDE) so I don't think bringing a dependency is worth the hassle.

Reply:
> Do you know if it does the translating or we still need to do it ourselves?

I don't know, sorry.

> TBF, this code already works (as it's taken from KDE) so I don't think bringing a dependency is worth the hassle.

That might be, just an option.
We would probably want to use that library in Plasma anyway, just something to keep in mind for the future.